test/test_corpus.cc

```diff
@@ -605,120 +605,130 @@ TEST_P(LSPTest, All) {
+
+                        if (!get_if<JSONNullObject>(&receivedCodeActionResponse)) {
+                            UnorderedMap<string, unique_ptr<CodeAction>> receivedCodeActionsByTitle;
+                            auto &receivedCodeActions = get<vector<unique_ptr<CodeAction>>>(receivedCodeActionResponse);
+                            for (auto &codeAction : receivedCodeActions) {
+                                bool codeActionTitleUnique = receivedCodeActionsByTitle.find(codeAction->title) ==
+                                                             receivedCodeActionsByTitle.end();
+                                EXPECT_TRUE(codeActionTitleUnique)
+                                    << "Found code action with duplicate title: " << codeAction->title;
+
+                                if (codeActionTitleUnique) {
+                                    receivedCodeActionsByTitle[codeAction->title] = move(codeAction);
+                                }
                             }
 
-                            // Ensure we received a code action matching the assertion.
-                            auto it2 = receivedCodeActionsByTitle.find(codeActionAssertion->title);
-                            EXPECT_NE(it2, receivedCodeActionsByTitle.end())
-                                << fmt::format("Did not receive code action matching assertion `{}` for error `{}`...",
-                                               codeActionAssertion->toString(), error->toString());
-
-                            // Ensure that the received code action applies correctly.
-                            if (it2 != receivedCodeActionsByTitle.end()) {
-                                auto codeAction = move(it2->second);
-                                codeActionAssertion->check(test.sourceFileContents, codeAction, test.testName, fileUri);
-
-                                // Some bookkeeping to make surfacing errors re. extra/insufficient
-                                // apply-code-action annotations easier.
-                                receivedCodeActionsByTitle.erase(it2);
-                                matchedCodeActionAssertions.emplace_back(*it);
-                                it = applyCodeActionAssertions.erase(it);
-                            } else {
-                                ++it;
+                            u4 receivedCodeActionsCount = receivedCodeActionsByTitle.size();
+                            vector<shared_ptr<ApplyCodeActionAssertion>> matchedCodeActionAssertions;
+
+                            // Test code action assertions matching the range of this error.
+                            auto it = applyCodeActionAssertions.begin();
+                            while (it != applyCodeActionAssertions.end()) {
+                                auto codeActionAssertion = it->get();
+                                if (cmpRanges(*codeActionAssertion->range, *error->range) != 0) {
```

Comment:
Do you still want the ranges to match exactly, or should it just match the region now that code actions works even if the query isn't an exact match?